packages/blocks/src/api/factory.js

```diff
@@ -345,13 +398,15 @@ export function switchToBlockType( blocks, name ) {
 	}
 
 	let transformationResults;
+
 	if ( transformation.isMultiBlock ) {
 		transformationResults = transformation.transform(
 			blocksArray.map( ( currentBlock ) => currentBlock.attributes ),
-			blocksArray.map( ( currentBlock ) => currentBlock.innerBlocks )
+			blocksArray.map( ( currentBlock ) => currentBlock.innerBlocks ),
+			blocksArray.map( ( currentBlock ) => currentBlock.name ),
 		);
 	} else {
-		transformationResults = transformation.transform( firstBlock.attributes, firstBlock.innerBlocks );
+		transformationResults = transformation.transform( firstBlock.attributes, firstBlock.innerBlocks, firstBlock.name );
```

Comment:
Oof. I really wish the original implementation of this API had passed the blocks directly, rather than the array of attributes. This mistake has forced these iterations, which I find incoherent to follow.

I wonder if it's still too late to try to correct course here, either:

- Creating a separate function with the desirable arguments signature (e.g. `transformBlocks`)
- Use (abuse?) characteristics of a function to "sniff" opt-in to this support (progressively support)

Specific to the second point, I actually think the use-case is well-satisfied by [generator functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*) anyways. I could imagine a function being implemented like:

```js
function* transform( blocks ) { 
	for ( const block of blocks ) {
		yield { ...block, name: 'core/changed' };
	}
}
```

Or the specific need here:

```js
function* transform( blocks ) {
	// Avoid transforming a single `core/group` Block
	if ( blocks.length === 1 && blocks[ 0 ].name === 'core/group' ) {
		return;
	}

	const alignments = [ 'wide', 'full' ];

	const widestAlignment = blocks.reduce( ( result, block ) => {
		// Determines the widest setting of all the blocks to be grouped
		const { align } = block.attributes;
		return alignments.indexOf( align ) > alignments.indexOf( result ) ? align : result;
	}, undefined );

	yield createBlock( 'core/group', {
		align: widestAlignment,
	}, blocks );
}
```

And from the code here, detecting `transform` as a generator function:

- https://stackoverflow.com/a/22610403
- https://stackoverflow.com/a/21559723

Reply:
Do we really want to use this PR to introduce generators support into the transforms API? I mean, this is potentially a subject that would allow us to solve async transforms as well (which is not great at the moment) and also raise the question about whether we could allow access to editor settings from the transforms.

I think it's a very big subject and I wonder if we should solve it more holistically. (and probably separately).

Reply:
...I read this as I have just pushed the code to implement this 😢 

Options:

1. As per my _original_ implementation, we pass a 3rd argument to the function which gives us access to the name. We then solve wider questions in another PR.

2. We create a new method `transformBlock` method which if provided takes precedence over `transform` and is passed the _entire_ block as an argument.

3. We leave the generator approach I've implemented in place.

Reply:
**Update**: [it was decided](https://wordpress.slack.com/archives/C02QB2JS7/p1559570594126300) to add a new `apply` method option and soft deprecate the `transform` method.

Reply:
This has now been added.